event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+	     (declare (ignorable parameters))
+	     ,(arguments args
+			 `(let ((res (progn ,@body)))
+			    (write! (make-instance 'response
+						   :keep-alive? t :content-type "text/event-stream")
+				    sock)
+			    (write! (make-instance 'sse :data (or res "Listening...")) sock)
+			    (force-output (socket-stream sock))))))
+```
+
+	(defmacro assert-http (assertion)
+	  `(unless ,assertion
+	     (error (make-instance 'http-assertion-error :assertion ',assertion))))
+
+`assert-http` is a macro that creates the boilerplate code we need in error cases. It expands into a check of the given assertion, throws an `http-assertion-error` if it fails, and packs the original assertion along in that event. 
+
+```lisp
+	(defmacro assert-http (assertion)
+	  `(unless ,assertion
+	     (error (make-instance 'http-assertion-error :assertion ',assertion))))
+```
+
+### HTTP "Types"
+
+In the previous section, we briefly touched on three expressions that we're using to implement our HTTP type validation system: `arg-exp`, `type-expression` and `type-assertion`. Once you understand those, there will be no magic left in our framework. We'll start with the easy one first.
+
+#### arg-exp
+
+`arg-exp` takes an argument symbol and creates an `aif` expression that checks for the presence of a parameter. 
+
+```lisp
+	(defun arg-exp (arg-sym)
+	  `(aif (cdr (assoc ,(->keyword arg-sym) parameters))
+		(uri-decode it)
+		(error (make-instance 'http-assertion-error :assertion ',arg-sym))))
+```
+
+Evaluating `arg-exp` on a symbol looks like:
+
+```lisp
+HOUSE> (arg-exp 'room)
+(AIF (CDR (ASSOC :ROOM PARAMETERS)) (URI-DECODE IT)
+     (ERROR (MAKE-INSTANCE 'HTTP-ASSERTION-ERROR :ASSERTION 'ROOM)))
+HOUSE> 
+```
+
+We've been using forms like `aif` and `awhen` without understanding how they work, so let's take some time to explore them now.
+
+Recall that Lisp code is itself represented as a tree. That's what the parentheses are for; they show us how leaves and branches fit together. If we step back to what we were doing in the previous section, `make-closing-handler` calls a function called `arguments` to generate part of the Lisp tree its constructing, which in turn calls some tree-manipulating helper functions, including `arg-exp`, to generate its return value. 
+
+That is, we've built a small system that takes a Lisp expression as input, produces a different Lisp expression as output. Possibly the simplest way of conceptualizing this is as a simple Common Lisp to Common Lisp compiler that is specialized to the problem at hand.
+
+A widely used classification of such compilers are called _anaphoric macros_ [^anaphors]. This term comes from the linguistic concept of an _anaphor_, which is the use of one word as a substitute for a group of words that preceded it. `aif` and `awhen` are anaphoric macros, and they're the only ones that I tend to often use. There are many more availabile in the [`anaphora` package](http://www.cliki.net/Anaphora).
```

Comment:
Can't find the other end of this `[^anaphors]` anywhere. Did it get edited out, or was that unintentional?